src/main/java/org/jvnet/hudson/test/IOUtil.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015 Red Hat, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jvnet.hudson.test;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.util.logging.Logger;
+
+public class IOUtil {
+
+    private static final Logger LOGGER = Logger.getLogger(IOUtil.class.getName());
+
+    /**
+     * Gives random available TCP port in the given range.
+     *
+     * @param from if &lt;=0 then default value 49152 is used
+     * @param to   if &lt;=0 then default value 65535 is used
+     */
+    public static int randomTcpPort(int from, int to){
+        from = (from <=0) ? 49152 : from;
+        to = (to <= 0) ? 65535 : to;
+
+
+        while(true){
+            int candidate = (int) ((Math.random() * (to-from)) + from);
+            if(isTcpPortFree(candidate)){
+                return candidate;
+            }
+            LOGGER.info(String.format("Port %s is in use", candidate));
+        }
+    }
+
+    /**
+     * Gives random available TCP port.
+     */
+    public static int randomTcpPort(){
+        return randomTcpPort(-1,-1);
+    }
+
+    public static boolean isTcpPortFree(int port){
+        try {
+            ServerSocket ss = new ServerSocket(port);
+            ss.close();
+            return true;
```

Comment:
I had actually thought about doing exactly this when first writing `RealJenkinsRule`. I avoided it because I thought that, besides the rare race condition, it would not actually work because the Linux kernel can decline to let you open a server socket for two minutes after closing one. Maybe that only applies to sockets that had actually accepted connections?

Reply:
the ATH does this and it still has random failures.  would guess this is trading one form of random failure for another.

Reply:
That's why i'm trying https://github.com/jenkinsci/winstone/pull/174 (+ downstreams pending)

Reply:
https://docs.oracle.com/javase/8/docs/api/java/net/Socket.html#setReuseAddress-boolean- sounds fine. While there would still be a race condition, it should be much less likely than in the status quo.

Reply:
sounds a bit complicated to me compared to use 0 port to get a free port such this one https://github.com/eclipse/jetty.project/blob/cd1f146867f66f37df73184a086f0e17828d70c8/tests/test-distribution/src/main/java/org/eclipse/jetty/tests/distribution/JettyHomeTester.java#L193
here you will eventually iterate on few used port before get a free whereas server port 0 just tell the os to give a free port. 
but as it seems to be a temporary solution until winstone patch...